src/Server.ts

```diff
@@ -0,0 +1,177 @@
+import { IClient, IMessage, MessageType } from './Interfaces';
+import * as dgram from 'dgram';
+
+export default class Server
+{
+    private _port: number = 8000;
+    private _socket: dgram.Socket;
+    private _clients: {
+        [id: number]: IClient;
+    } = {};
+    private _bound = false;
+
+    constructor(private _show_log = false)
+    {
+        this._setup_server();
+    }
+
+    listen(port?: number): void;
+    listen(callback?: (port: number) => void): void;
+    listen(port: number, callback?: (port: number) => void): void;
+    listen(port_or_callback?: number | ((port: number) => void), callback?: (port: number) => void): void
+    {
+        let cb: typeof callback;
+        this._port = 8000;
+        if (port_or_callback) {
+            if (typeof port_or_callback === 'function') {
+                cb = port_or_callback;
+            } else if (typeof port_or_callback === 'number') {
+                this._port = port_or_callback;
+                cb = callback;
+            }
+        }
+
+        this._socket.on('error', this._on_server_error.bind(this));
+        this._socket.on('message', this._on_server_message.bind(this));
+        this._socket.on('close', () => {
+            this._clients = {};
+        });
+        if (cb)
+        {
+            this._socket.on('listening', () =>
+            {
+                cb(this._port);
+            });
+        }
+
+        this._bound = true;
+        this._socket.bind(this._port);
+    }
+
+    shutdown(callback?: () => void): void {
+        if (!this._bound && callback)
+            return callback();
+        this._socket.close(() =>
+        {
+            this._setup_server();
+            if (callback)
+                callback();
+        });
+    }
+
+    private _setup_server()
+    {
+        this._socket = dgram.createSocket('udp4');
+        this._bound = false;
+    }
+
+    private _on_server_error(error: Error): void
+    {
+        console.error(`Server error:\n ${error.stack}`);
+        this._socket.close();
+    }
+
+    private _on_server_message(message: Buffer, remote_info: dgram.RemoteInfo): void
+    {
+        let parsed_message: IMessage;
+        try
+        {
+            parsed_message = JSON.parse(message.toString());
+        } catch
```

Comment:
Usually, with put `catch` on a new line as well
```suggestion
        }
        catch
```